GH-5148 Introduce "soft fail" for corrupt ValueStore

hmottestad opened the pull request:
GitHub issue resolved: #5148 <!-- add a Github issue number here, e.g #123. -->

Briefly describe the changes proposed in this PR:

<!-- short description of your change goes here -->

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

ebner commented:
Thanks for the fixes @hmottestad! I haven't had time yet to look into this, but I can contribute to the documentation if you want.

Following up on another comment from your review of my PR: adding a test for an artificially corrupted store may be difficult, I don't really know how to do this reliably. Perhaps it is possible to manipulate a NativeValue instance in the ValueStore to hold an illegal state that triggers the new soft fail functionality. Any other suggestions?

hmottestad commented:
I'm actually working on it right now. I have a test that sets each byte in the values.dat file to zero, one at a time, then tries to read all the data.

hmottestad commented on core/sail/nativerdf/src/test/java/org/eclipse/rdf4j/sail/nativerdf/NativeSailStoreCorruptionTest.java:
```diff
@@ -0,0 +1,224 @@
+	}
+
+	@Test
+	public void testCorruptValuesDatFileEntireValuesDatFile() throws IOException {
+		for (int i = 4; i < 437; i++) {
+			logger.debug("Corrupting byte at position " + i);
+			repo.shutDown();
+			restoreFile(dataDir, "values.dat");
+
+			overwriteByteInFile(new File(dataDir, "values.dat"), i, 0x0);
+
+			repo.init();
+
+			List<Statement> list = getStatements();
+			assertEquals(6, list.size());
+
+		}
+
+	}
```
This is actually the main test that tests every single byte, except the header bytes, in the values.dat file.

hmottestad commented on site/content/documentation/programming/repository.md:
````diff
@@ -98,6 +98,9 @@ import org.eclipse.rdf4j.sail.nativerdf.NativeStore;
 Repository repo = new SailRepository(new NativeStore());
 ```
 
+In the unlikely event of corruption the system property `org.eclipse.rdf4j.sail.nativerdf.softFailOnCorruptData` can be set to `true` to
+allow the NativeStore to output CorruptValue/CorruptIRI/CorruptIRIOrBNode/CorruptLiteral objects.
+
````
Documentation.

hmottestad commented:
@ebner I've added some test coverage now.

I think it might be best to also try to see what happens if the data is written to a trig or json-ld file, to test the very simple corrupt value implementations i added.

hmottestad commented on core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/btree/BTree.java:
```diff
@@ -292,6 +292,12 @@ public BTree(File dataDir, String filenamePrefix, int blockSize, int valueSize,
 			this.rootNodeID = buf.getInt();
 
+			if (rootNodeID == 0) {
+				if (nioFile.size() > blockSize) {
```
This is a somewhat brittle approach and it's really only relevant when using the soft fail property. So maybe this should be wrapped with the property too and throw something other than an illegal state exception.

aschwarte10 commented on core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java:
```diff
@@ -191,12 +200,62 @@ public NativeValue getValue(int id) throws IOException {
+				return (T) new CorruptIRIOrBNode(revision, id, ((CorruptValue) resultValue).getData());
+			}
+			logger.warn(
+					"Possible corrupt data consider setting the system property org.eclipse.rdf4j.sail.nativerdf.softFailOnCorruptDataAndRepairIndexes to true");
```
this message sounds a bit strange (though it is clear what it means. Maybe you can phrase it in two sentences, same in line 255 and line 629

aschwarte10 commented on core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/ValueStore.java:
```diff
@@ -534,49 +601,76 @@ private NativeValue data2value(int id, byte[] data) throws IOException {
+				logger.error("Soft fail on corrupt data: Invalid type {} for value with id {}", data[0], id);
+				return new CorruptUnknownValue(revision, id, data);
+			}
+			throw new SailException("Invalid type " + data[0] + " for value with id " + id
```
Maybe also here, two sentences.

aschwarte10 commented on core/sail/nativerdf/src/main/java/org/eclipse/rdf4j/sail/nativerdf/datastore/DataFile.java:
```diff
@@ -197,29 +203,50 @@ public byte[] getData(long offset) throws IOException {
-		// We have either managed to read enough data and can return the required subset of the data, or we have read
-		// too little so we need to execute another read to get the correct data.
-		if (dataLength <= data.length - 4) {
+		// If the data length is larger than 750MB, we are likely reading the wrong data. Probably data corruption.
```
how did you come up with 750MB? Is this a special size?

hmottestad replied:
It was what I experienced when testing.

I wouldn't think that a single literal would realistically be 750 MB. But I'll double check that the logic only applies if the system property is activated.

hmottestad replied:
I added some more context to the comment

aschwarte10 reviewed (commented):
I am too distant from the native store implementation and its index structures to give a proper review. From what I can tell with going through the PR the logic seems fine: I can follow the idea and the new setting is opt-in. 
Please see inline for minor (optional) comments on some messages
